Initial work on Mac patching

colincornaby opened the pull request:
This is kind of ugly but seems to work.

I'm using tgz for the Mac app bundle in since it's a directory. That should be ok in since the Mac client is the only one that will need to look at these manifests - but it would require the client now handle the tar format.

I don't know if there is a better way that doesn't involve tar'ing. I could store as separate files but the logic starts to get more complicated at knowing what to keep and what to delete if we want to get rid of the existing bundle entirely.

colincornaby commented:
I'm looking through to see if an alternate compression mechanism for directories would be better. Plasma doesn't include a TAR library but it does include a lot of other things. There might be an alternative that works better for Plasma.

Hoikas commented on urumanifest/commit.py:
```diff
@@ -34,6 +37,12 @@
 
 def _compress_asset(source_path: Path, output_path: Path):
     output_path.parent.mkdir(parents=True, exist_ok=True)
+    if isdir(source_path):
```
```suggestion
    if source_path.is_dir():
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -17,12 +17,15 @@
 import concurrent.futures
 from dataclasses import dataclass
 import functools
+from genericpath import isdir
```
```suggestion
```

This module doesn't seem to be documented.

Hoikas commented on urumanifest/commit.py:
```diff
@@ -34,6 +37,12 @@
 def _compress_asset(source_path: Path, output_path: Path):
     output_path.parent.mkdir(parents=True, exist_ok=True)
+    if isdir(source_path):
+        tar = tarfile.TarFile.open(output_path, "w:gz")
+        tar.add(source_path,  arcname="")
+        tar.close()
```
```suggestion
        with tarfile.TarFile.open(output_path, "w:gz") as tar:
            tar.add(source_path,  arcname="")
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -46,8 +55,14 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
     h = hashlib.md5()
-    with source_path.open("rb") as in_stream:
-        _io_loop(in_stream, h.update)
+    if source_path.is_dir:
```
```suggestion
    if source_path.is_dir():
```

This would previously always evaluate to truthiness because you were testing the bound method instead of calling it.

Hoikas commented on urumanifest/commit.py:
```diff
@@ -46,8 +55,14 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
-    with source_path.open("rb") as in_stream:
-        _io_loop(in_stream, h.update)
+    if source_path.is_dir:
+        for subdir, dirs, files in os.walk(source_path):
+            for f in files:
+                with Path(os.path.join(subdir, f)).open("rb") as in_stream:
+                    _io_loop(in_stream, h.update)
```
Use [`source_path.rglob("*")`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.rglob) instead.

Hoikas commented on urumanifest/constants.py:
```diff
@@ -62,7 +62,10 @@ class _manifests(NamedTuple):
 
     # Legacy -- to be deleted??? -- TransGaming Cider Wrapper (macOS)
-    "mac": _manifests(None, None, "macExternal"),
+    "mac": _manifests(None, None, "macExternal")
+    ,
```
```suggestion
    "mac": _manifests(None, None, "macExternal"),
```

Hoikas commented on urumanifest/github.py:
```diff
@@ -383,10 +385,17 @@ def _unpack_artifact(staging_path: Path, database: _WorkflowDatabase, rev: str,
+                if len(member_path.parts) >=  2 and member_path.parts[0] == "client":
+                    is_mac_app_bundle = member_path.parts[1].endswith(".app")
+                    if not i.is_dir() and len(member_path.parts) ==  2:
+                        yield (Path(i.filename).name, i, Path(i.filename).name)
```
```suggestion
                        yield (member_path.name, i, member_path.name)
```

Hoikas replied:
I would prefer to see this return a `NamedTuple` so I can know what exactly I'm dealing with in terms of return values.

Hoikas commented on urumanifest/github.py:
```diff
@@ -383,10 +385,17 @@ def _unpack_artifact(staging_path: Path, database: _WorkflowDatabase, rev: str,
+                    if not i.is_dir() and len(member_path.parts) ==  2:
+                        yield (Path(i.filename).name, i, Path(i.filename).name)
+                    elif is_mac_app_bundle and not i.is_dir():
+                        path = i.filename[7:]
```
What is this magic number doing?

Hoikas commented on urumanifest/github.py:
```diff
@@ -417,6 +428,7 @@ def _unpack_member(output_path: Path, archive: zipfile.ZipFile, info: zipfile.Zi
                 outfile.write(buf)
 
 
+
```
```suggestion
```

Hoikas commented on urumanifest/utils.py:
```diff
@@ -20,7 +20,7 @@
 def check_python_version(py_exe, py_version=(2,7)):
     logging.debug(f"Checking Python interpreter version: {py_exe}")
-    if not py_exe or not py_exe.is_file():
+    if not py_exe:
```
?

colincornaby replied:
This should not have gone in. I was getting an exception here and it was quicker just to disable this check. I'll put it back and then we can talk over the issue.

Hoikas commented on urumanifest/github.py:
```diff
@@ -396,19 +405,21 @@ def iter_client_dir():
 
         gather_key = workflow_lut[name]
-        gather_package = { gather_key: [PurePosixPath(i.filename).name for i in desired_members] }
+        gather_package = { gather_key: [i[2] for i in desired_members] }
```
What is `i[2]`?

Hoikas commented on urumanifest/github.py:
```diff
@@ -396,19 +405,21 @@ def iter_client_dir():
     with archive.open(info, "r") as infile:
-        with output_path.joinpath(Path(info.filename).name).open("wb") as outfile:
+        full_path = output_path.joinpath(output_subpath)
+        Path(os.path.dirname(full_path)).mkdir(parents=True, exist_ok=True)
```
What is going on here? This looks like it's making a new directory in the CWD...

colincornaby replied:
It's unpacking the app bundle into the gather directory - but it needs to create any missing folders in the unpack path along the way.

Hoikas commented on urumanifest/github.py:
```diff
@@ -20,8 +20,10 @@
+import os
 import json
 from pathlib import Path, PurePosixPath
+from queue import Full
```
Is this being used?

Hoikas commented on urumanifest/github.py:
```diff
@@ -20,8 +20,10 @@
 import itertools
 import logging
 import math
+import os
```
I don't see the `os` module being used...

Hoikas commented on urumanifest/commit.py:
```diff
@@ -109,7 +124,12 @@ def on_compress(asset: manifest.ManifestEntry, future: concurrent.futures.Future
-            asset_output_path = output_path.joinpath(server_path).with_suffix(f"{server_path.suffix}.gz")
+
+            if source_asset.source_path.is_dir():
+                staged_asset.flags |= ManifestFlags.bundle
```
It feels like we should be flagging this earlier in the process. This function is responsible only for compressing the assets.

colincornaby replied:
Not sure where would be better. It's weird to have this in a compression stage - but the bundle flag is also a direct result of how the compression stage decides to compress. I could see maybe driving it in reverse - something else marks it as a bundle and then uses the flag to do an alternative compression. I'm just not sure which function would be better for setting the flag.

Hoikas replied:
The correct place to set this is dependencies.py in `find_client_dependencies()` subfunction `track_manifest_dependency()`. Windows prerequisite installers are already being marked there based off their presence in the `prereq` sections. We probably want for the presence in the mac sections to work the same.

colincornaby committed:
Apply suggestions from code review

Co-authored-by: Adam Johnson <AdamJohnso@gmail.com>

colincornaby committed:
Update urumanifest/github.py

Co-authored-by: Adam Johnson <AdamJohnso@gmail.com>

colincornaby committed:
Initial work on Mac patching

colincornaby committed:
OS is needed for pathname manipulation

colincornaby committed:
Removing magic number in bundle path setup

colincornaby committed:
Switching to rglob

colincornaby committed:
Converting to named tuple

colincornaby commented:
@Hoikas - I haven't started on matching client side changes yet. I think I'd like feedback on the compression mechanism first.

This PR uses a tarfile in since it's usually complementary for gzip. But that's probably going to require a new library be used in Plasma if our dependencies aren't pulling in a tar library already.

We could also open things up to a wider range of archive formats.

Or I can back off and try to get things working single file. Each file in the bundle gets zipped separately and the flag activates some smarts about treating the parent directory as a single unit. We'd lose some compression by not bundling all the files together but that would be the most compatible with the existing approach.

Hoikas commented on urumanifest/github.py:
```diff
@@ -396,19 +408,21 @@ def iter_client_dir():
     with archive.open(info, "r") as infile:
-        with output_path.joinpath(Path(info.filename).name).open("wb") as outfile:
+        full_path = output_path.joinpath(output_subpath)
+        Path(os.path.dirname(full_path)).mkdir(parents=True, exist_ok=True)
```
```suggestion
        full_path.parent.mkdir(parents=True, exist_ok=True)
```

Hoikas commented on urumanifest/github.py:
```diff
@@ -383,10 +386,19 @@ def _unpack_artifact(staging_path: Path, database: _WorkflowDatabase, rev: str,
+                        yield ArtifactInfo(path=member_path.name, zipinfo=i, name=member_path.name)
+                    elif is_mac_app_bundle and not i.is_dir():
+                        # remove "client/"
+                        path = i.filename[len(client_folder_name)+1:]
+                        yield ArtifactInfo(path=Path(path), zipinfo=i, name=member_path.parts[1])
```
```suggestion
                        path = member_path.relative_to(client_folder_name)
                        yield ArtifactInfo(path=path.name, zipinfo=i, name=member_path.parts[1])
```

I must admit that I'm a little confused as to why we need to remove this directory name here, though.

colincornaby replied:
The directory is multilevel so we need to recreate the directory structure and deposit everything in game_binaries. But we don't want to consider the top level client directory to be part of the directory tree being copied over.

The existing code avoided this problem because it only cared about the file name and didn't care about the path - as far as I know.

Hoikas replied:
Still looking for less manual string fiddling here.

Hoikas commented on urumanifest/github.py:
```diff
@@ -374,6 +375,8 @@ def nuke_recurse(nuke: Path):
     database.current_sha = rev
     database.valid = True
 
+ArtifactInfo = namedtuple('ArtifactInfo', 'path zipinfo name')
```
I'd prefer to see this as a typed named tuple: https://docs.python.org/3/library/typing.html#typing.NamedTuple

Hoikas commented on urumanifest/commit.py:
```diff
@@ -34,6 +35,11 @@
     output_path.parent.mkdir(parents=True, exist_ok=True)
+    if source_path.is_dir():
+        with tarfile.TarFile.open(output_path, "w:gz") as tar:
+            tar.add(source_path,  arcname="")
```
```suggestion
            tar.add(source_path, arcname="")
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -34,6 +35,11 @@
+    if source_path.is_dir():
+        with tarfile.TarFile.open(output_path, "w:gz") as tar:
+            tar.add(source_path,  arcname="")
+        h = hashlib.md5(open(output_path,'rb').read())
```
This needs to use `_io_loop` to prevent excessive memory consumption.

Hoikas reviewed (commented):
The only concern I have right now is that we're going to be vulnerable to the order that files are hashed inside the tarachive. Hopefully, that won't come back to bite us... I don't think we currently have a tar library in Plasma. I am unfamiliar with the details of the format, so I can't offer any guidance there.

colincornaby committed:
Feedback from code review

colincornaby commented:
> The only concern I have right now is that we're going to be vulnerable to the order that files are hashed inside the tarachive. Hopefully, that won't come back to bite us... I don't think we currently have a tar library in Plasma. I am unfamiliar with the details of the format, so I can't offer any guidance there.

(Still working through the other feedback.)

Order that files are MD5'd is a risk. We'll have to make sure all ends are sorting the file the same way.

Tar is a Unix archive format usually used with gzip. Strangely - libtar does not work on Windows. But - libarchive does and libarchive can deal with tar archives. But libarchive also supports a wide array of formats so if we wanted to pick something a bit newer we could.

For macOS specifically - libarchive is also included with the system. Although nothing is stopping us from statically linking on macOS - as we would have to do for the other platforms anyway.

colincornaby committed:
Finishing implementing code review feedback

colincornaby committed:
Syncing with latest Mac self patching changes

Hoikas commented on urumanifest/commit.py:
```diff
@@ -46,8 +54,13 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
+        bundle_name = source_path.stem
+        # Guess the executable path - we could get the exact name if we could unpack the Info.plist
+        # in a cross platform way.
+        source_path = source_path.joinpath(Path("Contents/MacOS/" + bundle_name))
```
```suggestion
        source_path = source_path.joinpath(Path("Contents/MacOS",  bundle_name))
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -46,8 +54,13 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
+        source_path = source_path.joinpath(Path("Contents/MacOS/" + bundle_name))
     with source_path.open("rb") as in_stream:
-        _io_loop(in_stream, h.update)
+         _io_loop(in_stream, h.update)
```
```suggestion
        _io_loop(in_stream, h.update)
```

Erroneous whitespace change?

colincornaby commented on urumanifest/commit.py:
```diff
@@ -34,6 +35,13 @@
+        with output_path.open("rb") as in_stream:
+            h = hashlib.md5()
+            _io_loop(in_stream, h.update)
+        return h.hexdigest(), output_path.stat().st_size
```
Note to myself: I still need to address permissions here.

colincornaby committed:
Apply suggestions from code review

Co-authored-by: Adam Johnson <AdamJohnso@gmail.com>

colincornaby commented on urumanifest/commit.py:
```diff
@@ -46,6 +54,11 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
     server_path, source_path = args
     # One day, we will not use such a vulnerable hashing algo...
     h = hashlib.md5()
+    if source_path.is_dir():
```
Something I don't like here: The flag that this is an app bundle seems to be encoded later and isn't available yet. So I'm checking to see if the asset is a directory which is kind of gross. Ideally I'd only be checking the bitmask for the bundle property.

colincornaby committed:
Fixing the script yielding the file name to unpack instead of the path into the bundle. Not sure why this code was written this way, should have never worked…

colincornaby committed:
Properly setting permissions of archived app bundles

colincornaby committed:
Cleaning up bundles adding multiple name entries

colincornaby commented:
Going to promote this to ready to review.

The only odd output that I think is present here is that we archive to top level of the bundle instead of the bundle itself. I.E. if you have plClient.app/Contents, the top level of the archive will have Contents and not plClient.app. Even though it's odd - I still think it's proper. The manifest contains the plClient.app destination name - and not encoding that level into the archive avoids duplication between the manifest and the archive, and avoids possible naming conflicts and corner cases. I.E. if the manifest said the name was UruExplorer.app but the manifest said plClient.app - thats a corner case I don't want to deal with.

colincornaby committed:
Adding Mac thin client manifests

colincornaby commented:
The bundle flag is not properly being carried through the script - I'll dig in and figure out why it isn't landing in the final manifest.

colincornaby committed:
Fixing bundle flag

Client path does not include trailing slash. Checking for a .app suffix instead.

Hoikas commented on urumanifest/commit.py:
```diff
@@ -46,6 +59,11 @@ def _hash_asset(args: Tuple[Path, Path]) -> Tuple[Path, str, int]:
+        bundle_name = source_path.stem
+        # Guess the executable path - we could get the exact name if we could unpack the Info.plist
+        # in a cross platform way.
+        source_path = source_path.joinpath(Path("Contents/MacOS",  bundle_name))
```
```suggestion
        source_path = source_path.joinpath("Contents", MacOS",  bundle_name)
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -109,7 +127,11 @@ def on_compress(asset: manifest.ManifestEntry, future: concurrent.futures.Future
             assert server_path.parent.name
-            asset_output_path = output_path.joinpath(server_path).with_suffix(f"{server_path.suffix}.gz")
+
+            if staged_asset.flags & ManifestFlags.bundle:
+                asset_output_path = output_path.joinpath(server_path).with_suffix(f"{server_path.suffix}.tgz")
+            else:
+                asset_output_path = output_path.joinpath(server_path).with_suffix(f"{server_path.suffix}.gz")
```
```suggestion
            output_suffix = "tgz" if staged_asset.flags & ManifestFlags.bundle else "gz"
            asset_output_path = output_path.joinpath(server_path).with_suffix(f"{server_path.suffix}.{output_suffix}")
```

Flatten the logic for easier understanding.

Hoikas commented on urumanifest/dependencies.py:
```diff
@@ -191,6 +191,7 @@ def track_client_dependency(client_path: Path, server_path: Path):
 
     def track_manifest_dependency(client_path: Path, server_path: Path, category: str, manifest_names):
         flags = ManifestFlags.installer if category in gather_installers else 0
+        flags |= ManifestFlags.bundle if client_path.suffix == ".app" else 0
```
It would be nice to have the bundle extension in constants.py instead of as a magic string at the very minimum. On the other hand, the rest of the script (i.e. the Windows support) doesn't try to use magic strings in the filename to identify what flags need to be applied. Instead, we use the the keys from the control json files to assign a category. Then, we apply flags from the category here. So, what we should really be doing is have a dict in the control json to represent the new mac bundle. github.py will need to be updated to generate that part of the fake control json.

colincornaby replied:
I can take a look at that.

Hoikas replied:
Looks like this is still outstanding?

Hoikas commented on urumanifest/github.py:
```diff
@@ -383,10 +389,19 @@ def _unpack_artifact(staging_path: Path, database: _WorkflowDatabase, rev: str,
                 member_path = PurePosixPath(i.filename)
-                if len(member_path.parts) == 2 and member_path.parts[0] == "client":
-                    yield i
+                client_folder_name = "client"
+                if len(member_path.parts) >=  2 and member_path.parts[0] == client_folder_name:
+                    is_mac_app_bundle = member_path.parts[1].endswith(".app")
+                    if not i.is_dir() and len(member_path.parts) ==  2:
+                        yield ArtifactInfo(path=member_path.name, zipinfo=i, name=member_path.name)
+                    elif is_mac_app_bundle and not i.is_dir():
+                        # remove "client/"
+                        path = member_path.relative_to(client_folder_name)
+                        yield ArtifactInfo(path=path, zipinfo=i, name=member_path.parts[1])
```
What is going on here?

colincornaby replied:
This is constructing paths to the bundle, but not the files in the bundle. For example:
`client/plClient.app/Contents/Info.plist`
Should only be added to the control json as:
`plClient.app`

Hoikas commented on urumanifest/constants.py:
```diff
@@ -63,6 +63,8 @@ class _manifests(NamedTuple):
 
     # Legacy -- to be deleted??? -- TransGaming Cider Wrapper (macOS)
     "mac": _manifests(None, None, "macExternal"),
+    "macInternal": _manifests("MacThinInternal", None, "MacInternal"),
+    "macExternal": _manifests("MacThinExternal", None, "MacExternal"),
```
Are these legacy manifests?

colincornaby replied:
macInternal and macExternal are current. Should they be declared as something else?

Hoikas replied:
They should probably be listed before the comment that says legacy manifests follow.

Hoikas commented on urumanifest/constants.py:
```diff
@@ -81,6 +83,8 @@ class _directorytuple(NamedTuple):
     "external64": _directorytuple("", "client/windows_amd64/external"),
     "internal": _directorytuple("", "client/windows_ia32/internal"),
     "internal64": _directorytuple("", "client/windows_amd64/internal"),
+    "macexternal": _directorytuple("", "client/mac/external"),
+    "macinternal": _directorytuple("", "client/mac/internal"),
```
The filename case doesn't match above. Also, just to make sure - we are assuming that mac clients are universal binaries?

colincornaby replied:
Yes, thats the assumption. It wouldn't really make sense to have it be a universal binary locally and then a processor specific binary.

That said - that will make things messy when Apple eventually gets around to cutting Intel support from their tools. But that eventuality could be dealt with like any other platform we discontinue support for.

dpogue replied:
Note that currently CI is not generating universal binaries (because of vcpkg limitations). In theory we could figure out a way to `lipo` them together into a universal binary when we attach them to the published release.

Hoikas replied:
Note to self: the `gather_lut` keys are expected to be all lowercase.

dpogue commented on urumanifest/constants.py:
```diff
@@ -153,4 +158,6 @@ class PyToolsResultCodes(enum.IntEnum):
     "plasma-windows-x64-external-release": "external64",
     "plasma-windows-x86-internal-release": "internal",
     "plasma-windows-x64-internal-release": "internal64",
+    "plasma-macos-x64-internal-release": "macInternal",
+    "plasma-macos-x64-external-release": "macExternal",
```
Just noting that these are Intel x86_64 binaries (which is probably the best option for the moment, but they are not universal because they do not include arm64)

colincornaby replied:
Yeah - that still needs to be done on the CI side. For now, I'm kind of assuming whatever the platforms should be is what is provided by CI.

colincornaby commented:
I have a structure that I'm working on to separate the Mac bundle - but I don't know if it really fits with how things are supposed to work.
```
{
  "macExternal": [
    "banner.png",
    "banner@2x.png",
    "resource.dat"
  ],
  "macBundle": [
    "UruExplorer.app"
  ]
}
```

This would de-associated the macBundle with macExternal. But I don't think a child map inside macExternal is supported?

Hoikas commented:
> But I don't think a child map inside macExternal is supported?

No, I don't think any of the code is equipped to handle dictionaries inside of dictionaries.

You may want to separate `macBundle` into `macBundleExternal` and `macBundleInternal` to allow shipping down both internal and external release apps.

colincornaby commented:
@Hoikas - That was going to be my next question. My assumption is somewhere deeper in the code I need to take those keys and pair them with a destination.

colincornaby committed:
Initial work on macBundleExternal/ Internal keys

colincornaby commented:
I have an initial version in that uses a custom Gather key for the Mac bundle. However - deeper sections of code are still doing a folder check.

colincornaby committed:
Elimating dir checks for bundle

Hoikas commented on urumanifest/github.py:
```diff
@@ -415,8 +416,13 @@ def iter_client_dir():
-        gather_package = { gather_key: list(set([i.name for i in desired_members])) }
+        desired_bundles = list(set([i.name for i in list(filter(lambda x: x.bundle == True, desired_members))]))
+        gather_package = { gather_key: desired_files }
+        if len(desired_bundles) > 0:
```
```suggestion
        if desired_bundles:
```

Hoikas commented on urumanifest/commit.py:
```diff
@@ -142,7 +142,8 @@ def on_compress(asset: manifest.ManifestEntry, future: concurrent.futures.Future
                                          source_asset.source_path,
-                                         asset_output_path)
+                                         asset_output_path,
+                                         staged_asset.flags &  ManifestFlags.bundle)
```
```suggestion
                                         staged_asset.flags & ManifestFlags.bundle)
```

Hoikas reviewed (commented):
I still need to do a complete rereview, but this is what I see for the two new commits.

colincornaby committed:
Apply suggestions from code review

Co-authored-by: Adam Johnson <AdamJohnso@gmail.com>

Hoikas commented on urumanifest/commit.py:
```diff
@@ -42,10 +55,15 @@ def _compress_asset(source_path: Path, output_path: Path):
+        bundle_name = source_path.stem
+        # Guess the executable path - we could get the exact name if we could unpack the Info.plist
+        # in a cross platform way.
+        source_path = source_path.joinpath("Contents", MacOS",  bundle_name)
```
```suggestion
        source_path = source_path.joinpath("Contents", MacOS", bundle_name)
```

Hoikas commented on urumanifest/github.py:
```diff
@@ -374,6 +375,12 @@ def nuke_recurse(nuke: Path):
     database.current_sha = rev
     database.valid = True
 
+class ArtifactInfo(NamedTuple):
+    path: Path
+    zipinfo: zipfile.ZipInfo
+    name: str
+    bundle: bool = False
+
```
```suggestion
class ArtifactInfo(NamedTuple):
    path: Path
    zipinfo: zipfile.ZipInfo
    name: str
    bundle: bool = False


```

PEP8 wants two blank lines after classes.

colincornaby committed:
Apply suggestions from code review

Co-authored-by: Adam Johnson <AdamJohnso@gmail.com>

colincornaby committed:
Fixing typo in string literal

colincornaby committed:
Eliminating manual folder extension check
